### PR TITLE
Bugfix/too long courses crash TRNG-1316

### PR DESCRIPTION
**Pull request by SimonTheSourcerer**
### Description
Our JSON course structure is too deep which leads to serialization errors. Easiest fix is listing the all steps in a big list at the beginning of the training JSON which removes depths. To make it backwards compatible I wrote a new serializer which is based on the old one.

**Fixes**: # (issue)
Our JSON course structure is too deep which leads to serialization errors.

### CODE REVIEW

**Review by Maxsiom — COMMENTED** [review]
A training course with 100 steps (~15.4 MB) still crashes my Editor (you cannot even start Unity anymore with that file in it). So this is only an improvement?

**Review by Maxsiom — DISMISSED** [review]
Tested manually with training courses up to 200 steps and crowded transitions. (Course file sizes are reduced a lot, I like that)
Worked perfectly when being **in editor** and **built**.

File: Runtime/Serialization/NewtonsoftJsonSerializer/ImprovedNewtonsoftJsonCourseSerializer.cs [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+using System.Runtime.Serialization;
+using System.Text;
+using Innoactive.Creator.Core.Configuration.Modes;
+using Innoactive.Creator.Core.Serialization.NewtonsoftJson;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+namespace Innoactive.Creator.Core.Serialization
+{
+    public class ImprovedNewtonsoftJsonCourseSerializer : NewtonsoftJsonCourseSerializer
```
Comment (Gusinuhe) [inline Runtime/Serialization/NewtonsoftJsonSerializer/ImprovedNewtonsoftJsonCourseSerializer.cs:12]: Missing docs

File: Runtime/Serialization/NewtonsoftJsonSerializer/ImprovedNewtonsoftJsonCourseSerializer.cs [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+    {
+        /// <inheritdoc/>
+        public override string Name { get; } = "Improved Newtonsoft Json Importer";
+
+        protected override int Version { get; } = 2;
+
+        public override ICourse CourseFromByteArray(byte[] data)
```
Comment (Gusinuhe) [inline Runtime/Serialization/NewtonsoftJsonSerializer/ImprovedNewtonsoftJsonCourseSerializer.cs:19]:
```suggestion
        /// <inheritdoc/>
        public override ICourse CourseFromByteArray(byte[] data)
```

File: Runtime/Serialization/NewtonsoftJsonSerializer/ImprovedNewtonsoftJsonCourseSerializer.cs [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+            if (version == 1)
+            {
+                return base.CourseFromByteArray(data);
+            }
+
+            CourseWrapper wrapper = Deserialize<CourseWrapper>(data, CourseSerializerSettings);
+            return wrapper.GetCourse();
+        }
+
+        public override byte[] CourseToByteArray(ICourse course)
```
Comment (Gusinuhe) [inline Runtime/Serialization/NewtonsoftJsonSerializer/ImprovedNewtonsoftJsonCourseSerializer.cs:35]:
```suggestion
        /// <inheritdoc/>
        public override byte[] CourseToByteArray(ICourse course)
```

File: Runtime/Serialization/NewtonsoftJsonSerializer/NewtonsoftJsonCourseSerializer.cs [hunk 4]
```diff
@@ -2,6 +2,7 @@
 using System.Collections.Generic;
 using System.Linq;
 using System.Text;
+using Boo.Lang.Runtime;
```
Comment (Gusinuhe) [inline Runtime/Serialization/NewtonsoftJsonSerializer/NewtonsoftJsonCourseSerializer.cs:5]: 🤔

Comment (Gusinuhe) [reply Runtime/Serialization/NewtonsoftJsonSerializer/NewtonsoftJsonCourseSerializer.cs]: Is this ignored on purpose?

File: Runtime/Serialization/NewtonsoftJsonSerializer/ImprovedNewtonsoftJsonCourseSerializer.cs (hunk 1 above)
Comment (Gusinuhe) [reply Runtime/Serialization/NewtonsoftJsonSerializer/ImprovedNewtonsoftJsonCourseSerializer.cs:15]: Is this ignored on purpose?

File: Runtime/Serialization/NewtonsoftJsonSerializer/ImprovedNewtonsoftJsonCourseSerializer.cs (hunk 2 above)
Comment (Gusinuhe) [reply Runtime/Serialization/NewtonsoftJsonSerializer/ImprovedNewtonsoftJsonCourseSerializer.cs:23]: Is this ignored on purpose?

File: Runtime/Serialization/NewtonsoftJsonSerializer/ImprovedNewtonsoftJsonCourseSerializer.cs (hunk 3 above)
Comment (Gusinuhe) [reply Runtime/Serialization/NewtonsoftJsonSerializer/ImprovedNewtonsoftJsonCourseSerializer.cs:40]: Is this ignored on purpose?

File: Runtime/Serialization/NewtonsoftJsonSerializer/NewtonsoftJsonCourseSerializer.cs (hunk 4 above)
Comment (SimonTheSourcerer) [reply Runtime/Serialization/NewtonsoftJsonSerializer/NewtonsoftJsonCourseSerializer.cs]: pushed the change minutes after the rereview request, but it's not in there anymore

**Review by Gusinuhe — APPROVED** [review]

**Pull request merged**